core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -149,6 +155,67 @@ class FileMessageSet private[kafka](@volatile var file: File,
   }
 
   /**
+   * Search forward for the message whose timestamp is greater than or equals to the target timestamp.
+   *
+   * The search will stop immediately when it sees a message in format version before 0.10.0. This is to avoid
+   * scanning the entire log when all the messages are still in old format.
+   *
+   * @param targetTimestamp The timestamp to search for.
+   * @param startingPosition The starting position to search.
+   * @return None, if no message exists at or after the starting position.
+   *         Some(TimestampOffset) with (largest_timestamp_checked, the_next_offset_to_read) otherwise.
+   */
+  def searchForTimestamp(targetTimestamp: Long, startingPosition: Int): Option[TimestampOffset] = {
+    var maxTimestampChecked = Message.NoTimestamp
+    var lastOffsetChecked = -1L
+    val messagesToSearch = read(startingPosition, sizeInBytes)
+    for (messageAndOffset <- messagesToSearch) {
```

Comment:
Hmm, I actually did that in my first patch. ```searchForOffset()``` and ```searchForTimestamp``` were sharing a ```scanAndHandleMessages()``` function that does a special iteration by only looking at the headers. You suggested maybe we can just use the iterator in the comments and I thought it was a good suggestion because of the following reasons:

1. search by timestamp is a very infrequent call.
2. unlike searching by offset, searching by timestamp do need to look into the payload if the message is a compressed message. So copy the value into the buffer seems reasonable.
3. Because we have the index, the number of bytes we read from the disk should be relatively small assuming the timestamps are not abnormal.

So it seems not worth duplicating the code of the iterator here?